### PR TITLE
Implement PauliStringPhasorGate

### LINKED ISSUE #1561: PauliStringPhasor should be a Gate
Certainly the concept of a gate that does this operation is well defined.  Having it around as an object may not be super valuable, except that I really do think having the concept of a gate is very useful for defining what is supported by a device.  So now if we have QC that support this phasor we cannot say whether it is supported or not without talking about what qubits it is operating on.

### PR DESCRIPTION
**Pull request by daxfohl**
BREAKING CHANGE: PauliStringPhasor no longer inherits from PauliStringGateOperation.

Implements PauliStringPhasor in terms of a GateOperation on a new class PauliStringPhasorGate.

Mostly involved moving existing functions from the operation to the gate, and then having the operation call those methods under the hood.

Closes #1561, xref #4683

@tanujkhattar

### CODE REVIEW

File: cirq-core/cirq/protocols/pauli_expansion_protocol.py [hunk 1]
```diff
@@ -86,6 +86,10 @@ def pauli_expansion(
         return default
 
     num_qubits = matrix.shape[0].bit_length() - 1
+
+    if num_qubits == 0:
+        return default
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/pauli_expansion_protocol.py:91]: Why is this needed? Please add explicit tests.

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 2]
```diff
@@ -58,8 +59,119 @@ def __init__(
...
+    def __str__(self) -> str:
+        if self.exponent_pos == -self.exponent_neg:
+            sign = '-' if self.exponent_pos < 0 else ''
+            exponent = str(abs(self.exponent_pos))
+            return f'exp({sign}iπ{exponent}*{self.pauli_string})'
+        return f'({self.pauli_string})**{self.exponent_relative}'
+
+
+@value.value_equality(approximate=True)
+class PauliStringPhasorGate(raw_types.Gate):
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:148]: Please add explicit tests for `PauliStringPhasorGate` class.

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 3]
```diff
@@ -58,8 +59,119 @@ def __init__(
...
+    def exponent_neg(self):
+        return self._pauli_gate.exponent_neg
+
+    @property
+    def exponent_pos(self):
+        return self._pauli_gate.exponent_pos
+
+    @property
+    def pauli_string(self):
+        return self._pauli_gate.pauli_string.on(*self.qubits)
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:82]: Why do this rather expensive operation instead of directly storing the created `pauli_string` in this class at construction? To keep things consistent?

Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:82]: In some sense, this is a breaking change because these properties were earlier mutable (`gate.exponent_neg=0.1` would update the field earlier) but are now immutable (will result in an `AttributeError` now). 


Also, more tests please!

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
A first round of high level comments. My main question is that whether we can fix these inconsistencies at a higher level; i.e. 

Q1: Can we make `PauliString` a `GateOperation` ?

Q2: Can we make `PauliStringGateOperation` a `GateOperation` or deprecate it after this change?

**Comment by daxfohl** [comment]
> Q1: Can we make PauliString a GateOperation ?

I didn't notice that it wasn't (it has a `gate` field). It would be nice if we could. I'll look into that separately. I don't think it would affect what we do in this PR, but would be good to verify.

> Q2: Can we make PauliStringGateOperation a GateOperation or deprecate it after this change?

I think we should deprecate it. It doesn't appear to be a useful abstraction, it just provides some convenience functions that should probably be module-level, if anything. That said, we may want an interface `ContainsPauliString` that could be used for anything that has a `pauli_string` property.

File: cirq-core/cirq/protocols/pauli_expansion_protocol.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/protocols/pauli_expansion_protocol.py:91]: @viathor Need some numpy foo. I tracked this down to a root cause, and it appears `hilbert_schmidt_inner_product` fails for 1x1 matrices.

The reason tests now fail otherwise (without the above workaround), whereas they passed before, is that PauliStringPhasor did not define a _pauli_expansion_ method. But now since it inherits GateOperation, it does. There's a consistent_protocols test that checks this for an empty pauli string. This was fine before because the expansion protocol was not implemented, but now it is implemented and gets checked. The unitary matrix it uses in this case is 1x1, which fails at the inner_product step.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 3 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:82]: Good catch. I can add a property setter as well.

To note, I think we should immediately deprecate them though (if we add them at all). IIUC mutable operations invalidate the contract of FrozenCircuit. cc @95-martin-orion

File: cirq-core/cirq/protocols/pauli_expansion_protocol.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/protocols/pauli_expansion_protocol.py:91]: @viathor nvm, the problem is actually in `operator_spaces.kron_bases`, where the initial array is missing a dimension. That's an easy fix. I'll fix that and add a test.

**Comment by daxfohl** [comment]
@tanujkhattar Here's a slightly different approach where we don't break the inheritance from `PauliStringGateOperation`. https://github.com/quantumlib/Cirq/compare/master...daxfohl:gates-psp2?expand=1. This would be safer, I think, and follows the precedent of PauliString, in that it doesn't inherit GateOperation, but still provides a gate. (As I mentioned above, to me *having a gate* is the important thing; it doesn't matter whether it inherits GateOperation or not).

It's a little more redundant, since it has to do all the protocol forwarding. But safer since it doesn't break the hierarchy. I do see a couple places in the code where `isinstance(PauliStringGateOperation)` is used. So it's probably better to be safe here. If we want to deprecate `PauliStringGateOperation` later and change the hierarchy, we can do that separately. But I don't see a particular need to rush that.

(Ideally we could make `PauliStringGateOperation` inherit `GateOperation` and have the best of both worlds. But I can't think of a way to do that since it is not guaranteed to have a gate. Unless we give it a dummy `NoOpGate` if nothing else is provided.)

What do you think? Should we go with this alternate approach?

**Comment by daxfohl** [comment]
I went ahead and merged in the approach mentioned above, to ensure backwards compatibility with `PauliStringGateOperation` is preserved. It's a safer, smaller, simpler change. If we want to deprecate `PauliStringGateOperation` later we can revisit that as a separate effort.

**Comment by daxfohl** [comment]
Reverted this to bring back the hard break from PauliStringGateOperation as we'd discussed in cirq-sync noted above. @tanujkhattar

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 4]
```diff
@@ -58,138 +60,278 @@ def __init__(
...
         """
-        if pauli_string.coefficient == -1:
-            pauli_string = -pauli_string
-            exponent_pos, exponent_neg = exponent_neg, exponent_pos
+        gate = PauliStringPhasorGate(
+            dps.DensePauliString(pauli_string.values(), coefficient=pauli_string.coefficient),
+            exponent_neg=exponent_neg,
+            exponent_pos=exponent_pos,
+        )
+        super().__init__(gate, tuple(pauli_string))
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:70]: Please replace `tuple(pauli_string)` with `pauli_string.qubits`. The former is very hard to read.

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 5]
```diff
@@ -58,138 +60,278 @@ def __init__(
...
+        self._pauli_gate = gate
+        self._pauli_string = gate.dense_pauli_string.on(*self.qubits)
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:72]: Why is this not just `self._pauli_string = pauli_string` ?

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 6]
```diff
@@ -12,27 +12,29 @@
...
+    clifford_gate,
 )
 
 if TYPE_CHECKING:
     import cirq
 
 
 @value.value_equality(approximate=True)
-class PauliStringPhasor(pauli_string_raw_types.PauliStringGateOperation):
+class PauliStringPhasor(gate_operation.GateOperation):
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:37]: It looks to me that the main blocker in completely deprecating `PauliStringPhasor` operation and replacing all uses with `PauliStringPhaseGate` + the automatic gate operation would be 
- A bunch of convenience properties like `exponent_neg`, `exponent_pos` etc.; which currently exist on the operation and would later exist on the gate. 
- `pass_operations_over` method which modifies the underlying pauli string and hence would need to exist on the operation and not on the gate. 


I see very few uses of the `pass_operations_over` method in cirq and would argue in favour of deprecating it from the operation. Ideally, I would only want to have `PauliStringPhasorGate` and deprecate `PauliStringPhasor` operation. Let's discuss this again in the cirq sync today to see what others think about it.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 6 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:37]: I think that would be best, but also that it should be a separate PR.

Most of what breaks if we remove PauliStringPhasor is in contrib/pauli_string. I'm pretty sure everything can be patched up nicely to work with gate+qubits, but there's a fair amount of patching to do. Whereas the current PR is self-contained.

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 7]
```diff
@@ -58,138 +60,278 @@ def __init__(
...
-        to_z_ops = op_tree.freeze_op_tree(self.pauli_string.to_z_basis_ops())
+        to_z_ops = op_tree.freeze_op_tree(self._to_z_basis_ops(qubits))
         xor_decomp = tuple(xor_nonlocal_decompose(qubits, any_qubit))
         yield to_z_ops
         yield xor_decomp
 
-        if self.exponent_neg:
-            yield pauli_gates.Z(any_qubit) ** self.exponent_neg
-        if self.exponent_pos:
+        if self._exponent_neg:
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:278]: Use `self.exponent_neg` property instead of the internal variable `self._exponent_neg`. Here and elsewhere for all properties.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 5 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:72]: The normalization code is in the gate now. If we just use the raw input then it may not be normalized.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 6 above)
Comment (tanujkhattar) [reply cirq-core/cirq/ops/pauli_string_phasor.py:37]: Another alternative could be to define a private `_PauliStringPhasorOperation` which extends `GateOperation` and defines the additional convenience methods like `pass_operations_over`. 

What I do not want to is to continue supporting both public APIs:  `PauliStringPhasor` (an operation) and `PauliStringPhasorGate` (a gate). The two other places where we currently have this structure (i.e. both gate and operations for the same item are part of the public API) is `ControlledGate` & `ControlledOperation` and `DensePauliString` and `PauliString`. 

I'd like to avoid creating more such cases unless we see a strong use case. LMK your thoughts on this.

Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:37]: I'm in favor of deprecating and I'd strongly prefer deprecating it rather than creating a private thing that inherits GateOperation. But still, as a subsequent PR to this one.

Alternatives to consider for `pass_operations_over`: we could plop a `pass_operations_over` onto GateOperation, we could make it a protocol, or we could just move that logic to `contrib/pauli_string` rather than being part of the gate/operation itself.

Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:37]: Per cirq sync, we agreed it's best to deprecate `PauliStringPhasor` as a separate PR, so we can consider an approach with this and `PauliString` at the same time.

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 8]
```diff
@@ -58,26 +60,62 @@ def __init__(
...
         Raises:
-            ValueError: If the given pauli string does not have eignevalues +1
-                or -1.
+            ValueError: If coefficient is not 1 or -1.
         """
-        if pauli_string.coefficient == -1:
-            pauli_string = -pauli_string
-            exponent_pos, exponent_neg = exponent_neg, exponent_pos
+        gate = PauliStringPhasorGate(
+            dps.DensePauliString(pauli_string.values(), coefficient=pauli_string.coefficient),
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:66]:
```suggestion
            pauli_string.dense(pauli_string.qubits),
```

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 9]
```diff
@@ -58,26 +60,62 @@ def __init__(
...
+    @deprecated(
+        deadline="v0.15",
+        fix="The mutators of this class are deprecated, instantiate a new object instead.",
+    )
+    def pauli_string(self, pauli_string):
+        # coverage: ignore
+        self._pauli_string = pauli_string
+        self._pauli_gate._dense_pauli_string = dps.DensePauliString(
+            pauli_string.values(), coefficient=pauli_string.coefficient
+        )
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:114]:
```suggestion
        self._gate._dense_pauli_string = pauli_string.dense(pauli_string.qubits)
```

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 10]
```diff
@@ -58,26 +60,62 @@ def __init__(
...
+    @property
+    def pauli_string(self):
+        return self._pauli_string
+
+    @pauli_string.setter  # type: ignore
+    @deprecated(
+        deadline="v0.15",
+        fix="The mutators of this class are deprecated, instantiate a new object instead.",
+    )
+    def pauli_string(self, pauli_string):
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:109]: Ideally, `self.qubits` can also potentially change when updating the underlying pauli string. Big +1 on deprecating this but.

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 11]
```diff
@@ -58,26 +60,62 @@ def __init__(
...
-        if pauli_string.coefficient == -1:
-            pauli_string = -pauli_string
-            exponent_pos, exponent_neg = exponent_neg, exponent_pos
+        gate = PauliStringPhasorGate(
+            dps.DensePauliString(pauli_string.values(), coefficient=pauli_string.coefficient),
+            exponent_neg=exponent_neg,
+            exponent_pos=exponent_pos,
+        )
+        super().__init__(gate, pauli_string.qubits)
+        self._pauli_gate = gate
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:71]: We don't need to store an explicit `self._pauli_gate` because we are already storing the same pauli gate as `self._gate` because of deriving from `GateOperation` and calling `super().__init__`.

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 12]
```diff
@@ -58,26 +60,62 @@ def __init__(
...
+        # coverage: ignore
+        self._pauli_string = pauli_string
+        self._pauli_gate._dense_pauli_string = dps.DensePauliString(
+            pauli_string.values(), coefficient=pauli_string.coefficient
+        )
 
     @property
     def exponent_relative(self) -> Union[int, float, sympy.Basic]:
-        return value.canonicalize_half_turns(self.exponent_neg - self.exponent_pos)
+        return self._pauli_gate.exponent_relative
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:118]:
```suggestion
        return self.gate.exponent_relative
```

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 13]
```diff
@@ -88,9 +126,7 @@ def _value_equality_values_(self):
 
     def equal_up_to_global_phase(self, other):
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:127]: This is not correct as you are not checking for equality of qubits in here (only checking for equality of underlying gates). Qubit equality was implicitly checked earlier when comparing `pauli_string`s.

We could have used the implicit version defined in `GateOperation` if it used `self._group_interchangeable_qubits` to compare the qubits instead of directly using `self.qubits` (see #4766)

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 14]
```diff
@@ -58,26 +60,62 @@ def __init__(
...
+        self._pauli_gate._dense_pauli_string = dps.DensePauliString(
+            pauli_string.values(), coefficient=pauli_string.coefficient
+        )
 
     @property
     def exponent_relative(self) -> Union[int, float, sympy.Basic]:
-        return value.canonicalize_half_turns(self.exponent_neg - self.exponent_pos)
+        return self._pauli_gate.exponent_relative
 
     def _value_equality_values_(self):
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:120]: `_value_equality_values_` not needed as it'll be derived from `GateOperation` and `self.pauli_string` is redundant if we consider `self.gate` and `self.qubits`, which would automatically get considered in `GateOperation`

Comment (tanujkhattar) [reply cirq-core/cirq/ops/pauli_string_phasor.py:120]: We'd have to define a `qubit_index_to_equivalence_group_key` on the `PauliStringPhasorGate` for the implicit equality to work correctly. I leave the choice up to you on whether to use the implicit version or leave it as is.

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 15]
```diff
@@ -117,15 +146,131 @@ def merged_with(self, op: 'PauliStringPhasor') -> 'PauliStringPhasor':
...
+            ValueError: If coefficient is not 1 or -1.
+        """
+        if dense_pauli_string.coefficient == -1:
+            dense_pauli_string = -dense_pauli_string
+            exponent_pos, exponent_neg = exponent_neg, exponent_pos
+
+        if dense_pauli_string.coefficient != 1:
+            raise ValueError(
+                "Given PauliString doesn't have +1 and -1 eigenvalues. "
+                "pauli_string.coefficient must be 1 or -1."
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:215]: s/pauli_string/dense_pauli_string

File: cirq-core/cirq/ops/pauli_string_phasor.py [hunk 16]
```diff
@@ -117,15 +146,131 @@ def merged_with(self, op: 'PauliStringPhasor') -> 'PauliStringPhasor':
...
         return not self._is_parameterized_()
 
-    def _decompose_(self) -> 'cirq.OP_TREE':
-        if len(self.pauli_string) <= 0:
+    def _to_z_basis_ops(self, qubits: Sequence['cirq.Qid']) -> Iterator[raw_types.Operation]:
+        """Returns operations to convert the qubits to the computational basis."""
+        for i in range(len(self.dense_pauli_string)):
+            yield clifford_gate.SingleQubitCliffordGate.from_single_map(
+                {self.dense_pauli_string[i]: (pauli_gates.Z, False)}
+            )(qubits[i])
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/pauli_string_phasor.py:267]: Use `self.dense_pauli_string.on(*qubits).to_z_basis_ops()` instead of re-writing the method here.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Looking pretty good. Left a final round of comments.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 11 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:71]: It was to avoid casting everywhere, but I can change it.

(Ultimately we should probably make GateOperation generic to help in these cases, but outside the scope of this PR).

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 14 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:120]: Since PauliStringPhasor itself is likely to be removed at some point, I'll leave the additional complexity there.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 13 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py]: Nice catch!

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 14 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:120]: Or actually this is going to have to be done regardless if we move to GateOperation, so I'll implement that function.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 13 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py]: Will implement the equivalent qubits function instead.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 14 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:119]: This didn't work for me, as it still returned not equal. Reverting to the existing method.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 13 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py]: I tried, but it didn't seem to work. We can revisit this later if desired.

File: cirq-core/cirq/ops/pauli_string_phasor.py (hunk 14 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/pauli_string_phasor.py:119]: I see, interchangeable qubits isn't enough here. `X(0)*Y(1)` ~ `Y(1)*X(0)`, but you can't just interchange qubits you have to interchange the paulis too. Interchangeable qubits is sufficient for single-pauli-type gates like `X(0)*X(1)` but not mixed ones.

File: cirq-core/cirq/protocols/json_test_data/spec.py [hunk 17]
```diff
@@ -53,6 +53,7 @@
         'ParallelGateFamily',
         'PauliInteractionGate',
         'PauliStringPhasor',
+        'PauliStringPhasorGate',
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/json_test_data/spec.py:56]: Since `dense_pauli_string` is serializable, it'll be trivial to make `PauliStringPhasorGate` serializable so let's do that in this PR and add the corresponding `.json` and `.repr` files.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Final round of comments. Thanks for the iterations!

File: cirq-core/cirq/protocols/json_test_data/spec.py (hunk 17 above)
Comment (tanujkhattar) [reply cirq-core/cirq/protocols/json_test_data/spec.py:51]: Also make `PauliStringPhasor` serializable

**Review by tanujkhattar — APPROVED** [review]
LGTM
